// @adonisjs/http-server/build/src/Cookie/Client/index.js
"use strict";
/*
 * @adonisjs/http-server
 *
 * (c) Harminder Virk <virk@adonisjs.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CookieClient = void 0;
/// <reference path="../../../adonis-typings/index.ts" />
const set_cookie_parser_1 = __importDefault(require("set-cookie-parser"));
const PlainCookie = __importStar(require("../Drivers/Plain"));
const SignedCookie = __importStar(require("../Drivers/Signed"));
const EncryptedCookie = __importStar(require("../Drivers/Encrypted"));
/**
 * Cookie client exposes the API to parse/set AdonisJS
 * cookies as a client.
 */
class CookieClient {
    constructor(encryption) {
        this.encryption = encryption;
    }
    /**
     * Encrypt a key value pair to be sent in the cookie header
     */
    encrypt(key, value) {
        return EncryptedCookie.pack(key, value, this.encryption);
    }
    /**
     * Sign a key value pair to be sent in the cookie header
     */
    sign(key, value) {
        return SignedCookie.pack(key, value, this.encryption);
    }
    /**
     * Encode a key value pair to be sent in the cookie header
     */
    encode(_, value) {
        return PlainCookie.pack(value);
    }
    /**
     * Unsign a signed cookie value
     */
    unsign(key, value) {
        return SignedCookie.canUnpack(value) ? SignedCookie.unpack(key, value, this.encryption) : null;
    }
    /**
     * Decrypt an encrypted cookie value
     */
    decrypt(key, value) {
        return EncryptedCookie.canUnpack(value)
            ? EncryptedCookie.unpack(key, value, this.encryption)
            : null;
    }
    /**
     * Decode an encoded cookie value
     */
    decode(_, value) {
        return PlainCookie.canUnpack(value) ? PlainCookie.unpack(value) : null;
    }
    /**
     * Parses the set-cookie header and returns an
     * array of parsed cookies
     */
    parse(setCookieHeader) {
        const cookies = (0, set_cookie_parser_1.default)(setCookieHeader);
        return cookies.map((cookie) => {
            cookie.encrypted = false;
            cookie.signed = false;
            const value = cookie.value;
            /**
             * Unsign signed cookie
             */
            if (SignedCookie.canUnpack(value)) {
                cookie.value = SignedCookie.unpack(cookie.name, value, this.encryption);
                cookie.signed = true;
                return cookie;
            }
            /**
             * Decrypted encrypted cookie
             */
            if (EncryptedCookie.canUnpack(value)) {
                cookie.value = EncryptedCookie.unpack(cookie.name, value, this.encryption);
                cookie.encrypted = true;
                return cookie;
            }
            /**
             * Decode encoded cookie
             */
            if (PlainCookie.canUnpack(value)) {
                cookie.value = PlainCookie.unpack(value);
            }
            return cookie;
        });
    }
}
exports.CookieClient = CookieClient;
